### ISSUE TITLE
overclock question

### ISSUE THREAD

**Issue by negativeExponent** [libretro/nestopia#51]
the overclock logic is incorrect.

https://github.com/libretro/nestopia/blob/c24ffe8ef9c8d6b89f5a3d8b64105168a55e03d2/source/core/NstApu.cpp#L2382-L2390

overclockingIsSafe will always be true because regs.address is guaranteed to be at least have a value of 0xC000

https://github.com/libretro/nestopia/blob/c24ffe8ef9c8d6b89f5a3d8b64105168a55e03d2/source/core/NstApu.cpp#L2392-L2400

same here. regs.lengthCounter will always be at least 1.

**Comment by hizzlekizzle** [comment]
What does "safe" actually mean in this context?

**Comment by negativeExponent** [comment]
when DMC is latched, thats considered unsafe. It causes PCM audio channel to run faster than it should when overclocking. If this was following FCEUx ppu overclock logic, Reg2 and Reg3 writes should only check value of incoming data

https://github.com/libretro/libretro-fceumm/blob/59771792b0709f10c9bc7b408e41ae6df55d6dab/src/sound.c#L289-L302

**Comment by keithbowes** [comment]
Well, you should probably forward these concerns to the [upstream code](https://gitlab.com/jgemu/nestopia.git/).  I don't think anyone here does anything for the Nestopia backend but only for the libretro interface, so if this were fixed upstream, it would get merged into this core later.

**Comment by negativeExponent** [comment]
Upstream (if it's still even active) does not have overclock option, not has the libretro port. either case the condition is still incorrect and will always be true.

**Comment by keithbowes** [comment]
You're right.  I didn't think the core had anything that the upstream backend didn't.  I guess I was wrong.

Judging from the git logs, it seems that upstream did once have overclocking but it was considered buggy and was removed, but the project's originator readded the code specifically for the libretro core.

**Comment by carmiker** [comment]
Overclock did not survive the major accuracy improvements, and would have to be rewritten. I also continue to receive bug reports from people using overclocking, where I have to ask if they forgot to turn it off before testing if this is really an emulation bug or not. Every time, it is that they were using overclocking. If anyone wishes to overclock the NES, there are other emulators that do this better than Nestopia ever did.
